r2/service.go

```diff
@@ -59,86 +64,480 @@ func NewService(iOpts instrument.Options, rootPrefix string, store Store) server
 	return &service{iOpts: iOpts, store: store, rootPrefix: rootPrefix}
 }
 
+type payload struct {
+	data   []byte
+	status int
+}
+
+func newPayload(d []byte, status int) payload {
+	return payload{data: d, status: status}
+}
+
+func (p payload) write(w http.ResponseWriter) {
+	w.Header().Set("Content-Type", "application/json")
+	w.WriteHeader(p.status)
+	w.Write(p.data)
+}
+
+func writeAPIResponse(w http.ResponseWriter, code int, msg string) {
+	j, _ := json.Marshal(apiResponse{Code: code, Message: msg})
+	newPayload(j, code).write(w)
+}
+
+func handleError(w http.ResponseWriter, opError error) {
+	switch opError.(type) {
+	case ConflictError:
+		writeAPIResponse(w, http.StatusConflict, opError.Error())
+	case BadInputError:
+		writeAPIResponse(w, http.StatusBadRequest, opError.Error())
+	case VersionError:
+		writeAPIResponse(w, http.StatusConflict, opError.Error())
+	case NotFoundError:
+		writeAPIResponse(w, http.StatusNotFound, opError.Error())
+	default:
+		writeAPIResponse(w, http.StatusInternalServerError, opError.Error())
+	}
+}
+
+// TODO(dgromov): Make this return a list of validation errors
+func parseRequest(s interface{}, body io.ReadCloser) string {
+	if err := json.NewDecoder(body).Decode(s); err != nil {
+		return fmt.Sprintf("Malformed Json: %s", err.Error())
+	}
+
+	validate := validator.New()
```

Comment:
Is there a reason you have to do this validation manually (instead of relying on the validator/unmarshall interface)? if so, put a brief note about it